core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -557,15 +565,60 @@ trait EntityComponent {
       }
     }
 
-    def copyEntities(sourceWorkspaceContext: SlickWorkspaceContext, destWorkspaceContext: SlickWorkspaceContext, entityType: String, entityNames: Seq[String]): ReadWriteAction[TraversableOnce[Entity]] = {
-      getEntitySubtrees(sourceWorkspaceContext, entityType, entityNames).flatMap { entities =>
-        getCopyConflicts(destWorkspaceContext, entities).flatMap { conflicts =>
-          if (conflicts.isEmpty) {
-            cloneEntities(destWorkspaceContext, entities).map(_ => Seq.empty[Entity])
+    case class ConflictSubtree(entity: Entity, conflicts: Seq[Entity], allChildren: Seq[Entity]) {
+      def hasConflicts = conflicts.nonEmpty
+    }
+
+    def copyEntities(sourceWorkspaceContext: SlickWorkspaceContext, destWorkspaceContext: SlickWorkspaceContext, entityType: String, entityNames: Seq[String], linkExistingEntities: Boolean): ReadWriteAction[EntityCopyResponse] = {
+
+      def getAllAttrRefs(attrs: AttributeMap): Seq[AttributeEntityReference] = {
+        attrs.values.filter(_.isInstanceOf[AttributeEntityReference]).map(_.asInstanceOf[AttributeEntityReference]).toSeq
```

Comment:
I haven't looked closely yet but `filter` + `map` can often be `collect` instead